### PR TITLE
Add blog post for debugging slow bazel builds

### PR DESCRIPTION
**Pull request by maggie-lou**
I will wait to merge this for a week or 2 for some time between Brentley's post

### CODE REVIEW

File: website/blog/why-was-my-bazel-build-so-slow.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+image: /img/slow.png
+tags: [debugging, product, engineering]
+---
+
+The promise of Bazel is the promise of fast builds. So what do you do if your build was slow? Curse out the developer that convinced your company to migrate to Bazel? Of course not! You’d never do that, right? 🙂
+
+BuildBuddy is here to help. Bazel provides a lot of helpful information to debug slow builds, but it can be overwhelming to know where to look.
+
+Maybe you ran two builds and you’re surprised the second build ran so slowly, when you’d expected it to be cached. Or maybe you ran one build, and you just want to understand where all the time was spent. Here are a couple good places to start, with some real-world examples of why our own builds were slow.
+
```
Comment (iain-macdonald) [inline website/blog/why-was-my-bazel-build-so-slow.md:18]: Mind throwing a `<!-- truncate -->` tag in here somewhere so only a snippet shows up in the blog homepage?

File: website/blog/why-was-my-bazel-build-so-slow.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+author_image_url: https://avatars.githubusercontent.com/u/13951661
+image: /img/slow.png
+tags: [debugging, product, engineering]
+---
+
+The promise of Bazel is the promise of fast builds. So what do you do if your build was slow? Curse out the developer that convinced your company to migrate to Bazel? Of course not! You’d never do that, right? 🙂
+
+BuildBuddy is here to help. Bazel provides a lot of helpful information to debug slow builds, but it can be overwhelming to know where to look.
+
+Maybe you ran two builds and you’re surprised the second build ran so slowly, when you’d expected it to be cached. Or maybe you ran one build, and you just want to understand where all the time was spent. Here are a couple good places to start, with some real-world examples of why our own builds were slow.
```
Comment (iain-macdonald) [inline website/blog/why-was-my-bazel-build-so-slow.md:17]: More than "a couple" here :-)

File: website/blog/why-was-my-bazel-build-so-slow.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+This post frequently references the BuildBuddy UI. While most of these steps are
+possible without it, it will be a lot more work to find and parse the info from Bazel yourself.
```
Comment (iain-macdonald) [inline website/blog/why-was-my-bazel-build-so-slow.md:20]: s/Bazel/build event stream ?

File: website/blog/why-was-my-bazel-build-so-slow.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+- `--experimental_remote_cache_async`
+  - This flag can improve remote cache performance by enabling cache I/O in the background.
+- `--noremote_upload_local_results`
+  - This flag can reduce uploads by not uploading locally executed action outputs to the remote cache. This will reduce the cache hit rate for future runs, but can be desirable if upload speed is constrained (due to a poor network connection, for example).
+- `--remote_cache_compression`
+  - This flag can improve remote cache throughput by compressing cache blobs.
+- `--digest_function=BLAKE3`
+  - This flag can improve the performance of digest calculation of large files by using a faster hashing algorithm. This is available for Bazel 6.4+.
+- `--jobs`
+  - This flag can improve execution throughput by allowing more actions to execute in parallel.
```
Comment (iain-macdonald) [inline website/blog/why-was-my-bazel-build-so-slow.md:49]: Could you mention our recommended value for this flag, if we have one?

File: website/blog/why-was-my-bazel-build-so-slow.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+⭐ The details on the top of our UI are helpful for a quick sanity check here. If I’m comparing two builds, my first step will often be to compare the details on the top of the invocation pages and verify they’re exactly the same.
+
+For example, once I was inspecting a slow build, and realized I’d forgotten to set `--remote_download_minimal`. The build was downloading a lot more artifacts from the cache than usual, which explained the performance hit.
+
+![](../static/img/blog/ui-details.png)
+
+_The top invocation does not have –remote_download_minimal enabled, which could negatively impact performance._
```
Comment (iain-macdonald) [inline website/blog/why-was-my-bazel-build-so-slow.md:57]: Maybe worth recommending these are all set via a `.bazelrc` that's checked into git?

File: website/blog/why-was-my-bazel-build-so-slow.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+⭐ You may also want to consider how many threads are working on analysis / dependency fetching. Allocating more CPUs to the build and parallelizing it further may result in performance gains.
+
+In the example images included above, you can see that only two threads are working on analysis. Bazel tries to determine an appropriate number of threads based on the available CPU cores on your machine, but you can manually set it with the `--jobs` flag.
+
+## 3) Does Bazel have enough memory?
+
+Famously, Bazel loves memory. This section is short and sweet, but describes a common problem we’ve seen with our customers.
+
+⭐ If Bazel doesn’t have enough memory, it will evict parts of the analysis cache. As described in the previous section, this can significantly slow down builds. If you notice a lot of `MerkleTree.build` actions even when you’re reusing a warm Bazel cache or see heavy garbage collection activity in the timing profile, you may want to increase the amount of RAM available to Bazel by increasing the JVM heap size with `--host_jvm_args`.
```
Comment (iain-macdonald) [inline website/blog/why-was-my-bazel-build-so-slow.md:114]: Do we have a value we usually recommend, even if it's relative to the system properties? Or if not, maybe it's worth mentioning what the default is so it's obvious where to start?

File: website/blog/why-was-my-bazel-build-so-slow.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
+---
+slug: debugging-slow-bazel-builds
+title: Why is my Bazel build so slow?
+author: Maggie Lou
+author_title: "Software Engineer @ BuildBuddy"
+date: 2024-3-14 12:00:00
```
Comment (iain-macdonald) [inline website/blog/why-was-my-bazel-build-so-slow.md:6]: Mind updating the date when you merge?

**Review by iain-macdonald — CHANGES_REQUESTED** [review]
On one of my blog posts, Brandon recommended using webp images instead of pngs and that reduced the size significantly which is good for page load time and to reduce the size of our git repo. Would you mind converting the pngs to webps and overwriting the commit and force pushing so the pngs aren't in our repository too?

**Review by siggisim — APPROVED** [review]

**Pull request merged**